### PR TITLE
[mlir][Vector] Update v.contract -> v.outerproduct tests (NFC)

### PR DESCRIPTION
**Pull request by banach-space**
Re-orders tests in vector-contract-to-outerproduct-transforms.mlir so
that the file starts as follows:

  1. plain matmul
  2. plain matmul with scalable vectors
  3. masked matmul
  4. masked matmul with scalable vectors
  5. plain matmul with mixed types
  6. plain matmul with mixed types and scalable vectors

All of the above share the same indexing maps. This allowed to identify
one more duplicate. Following the cases above are examples with
different maps.

In addition, added extra comments to document the tests and to split
them into categories. There is also some extra reformatting to unify the
tests.

### CODE REVIEW

File: mlir/test/Dialect/Vector/vector-contract-to-outerproduct-transforms.mlir [hunk 1]
```diff
@@ -141,48 +92,66 @@ func.func @matmul(%arg0: vector<2x4xf32>,
...
-                          %arg2: vector<2x[3]xf32>) -> vector<2x[3]xf32> {
-  %0 = vector.contract #matmat_trait %arg0, %arg1, %arg2
+                           %arg1: vector<4x[3]xf32>,
+                           %arg2: vector<2x[3]xf32>) -> vector<2x[3]xf32> {
+  %0 = vector.contract #matmat_trait_0 %arg0, %arg1, %arg2
     : vector<2x4xf32>, vector<4x[3]xf32> into vector<2x[3]xf32>
   return %0 : vector<2x[3]xf32>
 }
 
-// CHECK-LABEL: func @matmul_0
```
Comment (MacDue) [inline mlir/test/Dialect/Vector/vector-contract-to-outerproduct-transforms.mlir:151]: Just checking: This is also deleting `matmul_0` and `matmul_0_scalable` as they (pretty much) duplicated the `masked_extract_contract4` tests?

File: mlir/test/Dialect/Vector/vector-contract-to-outerproduct-transforms.mlir [hunk 2]
```diff
@@ -1,15 +1,22 @@
...
-  indexing_maps = #matmat_accesses,
-  iterator_types = ["parallel", "parallel", "reduction"]
-}
+// NOTE - tests in this file are duplicated so that there's a version for
+//    * _fixed width_ and for _scalable_ vectors.
+// In order for the "vector.contract -> vector.outerproduct" patterns to work,
+// only the non-reduction dimension can be scalable (*). For Matmul operations
+// that is set to be the N dimension (i.e. rows of the output matrix), which
+// matches how matrix multiplication are normally implemented for e.g. 
+// Arm's SVE. However, making the M dimension scalable (i.e. columns of the
```
Comment (MacDue) [inline mlir/test/Dialect/Vector/vector-contract-to-outerproduct-transforms.mlir:9]: nit:
```suggestion
// for Arm SVE. However, making the M dimension scalable (i.e. columns of the
```

**Review by MacDue — APPROVED** [review]
Looks like a nice cleanup :+1:

File: mlir/test/Dialect/Vector/vector-contract-to-outerproduct-transforms.mlir (hunk 1 above)
Comment (banach-space) [reply mlir/test/Dialect/Vector/vector-contract-to-outerproduct-transforms.mlir:151]: Almost, so:
* `@masked_extract_contract4` -> `@masked_matmul` (for consistency),
* `@matmul_0` duplicates `@matmul`, so is deleted (the latter has more complete `CHECK` lines). The dimensions are a bit different, but it's the indexing maps that matter here, so I'm not concerned.

I could've done much better job documenting this in the summary, sorry. Also, only after a few iterations the duplication becomes apparent.

**Pull request merged**